{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "91ba513c-40c9-47df-96a4-8830d65021bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Ingredient:\n",
    "    def __init__(self, name, quantity):\n",
    "        self.name = name\n",
    "        self.quantity = quantity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a90177b5-c737-4c41-87af-b41f59f0d226",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['r']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['r', 'r']\n",
      "['r', 'r', 'r']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n",
      "Exception in Tkinter callback\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\tkinter\\__init__.py\", line 1948, in __call__\n",
      "    return self.func(*args)\n",
      "           ^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Admin\\AppData\\Local\\Temp\\ipykernel_10636\\614731485.py\", line 76, in update_ingredient\n",
      "    update_name_value = update_name.get()\n",
      "                        ^^^^^^^^^^^\n",
      "NameError: name 'update_name' is not defined\n"
     ]
    }
   ],
   "source": [
    "from tkinter import *\n",
    "\n",
    "# add module thông báo\n",
    "\n",
    "from tkinter import messagebox\n",
    "# tạo cửa sổ ứng dụng\n",
    "window = Tk()\n",
    "\n",
    "# setup thông tin cho window\n",
    "\n",
    "window.title(\"Ingredient Management System\")\n",
    "window.geometry('350x150')\n",
    "\n",
    "# tạo list để lưu trữ toàn bộ các objects trong khi vận hành\n",
    "list_ingredient = []\n",
    "\n",
    "# bước này làm sau khi thực hiện dán các widget lên window\n",
    "# mô tả các function để thực  hiện các hành động\n",
    "# các hành động là add, search, update, delete, reset\n",
    "# func add\n",
    "'''yêu cầu: thêm nguyên liệu mới vào list_ingredient.\n",
    "Mỗi ingredient sẽ đc  lưu trữ dưới dạng 1 object của class Ingredient.\n",
    "\n",
    "'''\n",
    "def add_ingredient():\n",
    "    # sau khi tạo class ở trên thì tiếp tục lấy giá trị từ các widget\n",
    "    name_value = name.get()\n",
    "    quantity_value = int(quantity.get())\n",
    "    \n",
    "    new_ingredient = Ingredient(name_value, quantity_value)\n",
    "    \n",
    "    list_ingredient.append(new_ingredient)\n",
    "\n",
    "    print([item.name for item in list_ingredient])\n",
    "\n",
    "    # hiện thị thông báo đã add thành công hay chưa\n",
    "    messagebox.showinfo('Add Ingredient', f\"{name_value} is added successfully!\")\n",
    "\n",
    "    # xoá các giá trị ở các entry\n",
    "    name.delete(0, END)# XOÁ TỪ ĐẦU ĐẾN CUỐI\n",
    "    quantity.delete(0, END)\n",
    "\n",
    "# buổi demo thứ 3 ngày 23.9, 3.1\n",
    "def search_ingredient():\n",
    "    search_name_value = search_name.get()\n",
    "\n",
    "    is_exist = False\n",
    "    \n",
    "    for ingredient in list_ingredient:\n",
    "        if ingredient.name == search_name_value:\n",
    "            messagebox.showinfo(\"Search Ingredient\", f'Name: {ingredient.name}, quantity: {ingredient.quantity}')\n",
    "            is_exist = True # đã tìm thấy nguyên liệu trong hệ thống\n",
    "\n",
    "    if is_exist == False: # không tìm thấy nguyên liệu nào trong suốt quá trình lặp\n",
    "        messagebox.showinfo(\"Search Ingredient\", 'Ingredient not found!')\n",
    "    \n",
    "    \n",
    "# def update_ingredient():\n",
    "#     name_value = name.get()\n",
    "#     quantity_value = quantity.get()\n",
    "    \n",
    "#     is_exist = False\n",
    "    \n",
    "#     for i in range(len(list_ingredient)):\n",
    "#         if list_ingredient[i].name == name_value:\n",
    "#             list_ingredient[i].quantity = quantity_value\n",
    "#             is_exist = True\n",
    "#             messagebox.showinfo(\"Update Ingredient\", f\"Update quantity of {name_value} successfully!\")\n",
    "#             break\n",
    "            \n",
    "#     if is_exist == False:\n",
    "#         messagebox.showinfo(\"Update Ingredient\", \"Ingredient not found!\")\n",
    "    \n",
    "\n",
    "def update_ingredient():\n",
    "    update_name_value = update_name.get()\n",
    "    update_quantity_value = update_quantity.get()\n",
    "    \n",
    "\n",
    "    is_exist = False\n",
    "    \n",
    "    for i in (len(list_ingredient)):\n",
    "        if list_ingredient[i].name == update_name_value:\n",
    "            list_ingredient[i].quantity == update_quantity_value\n",
    "            is_exist = True\n",
    "            messagebox.showinfo(\"Update Ingredient\", f\"Update quantity of {update_name_value} successfully!\")\n",
    "            break\n",
    "\n",
    "    if is_exist == False: # không tìm thấy nguyên liệu nào trong suốt quá trình lặp\n",
    "        messagebox.showinfo(\"Update Ingredient\", 'Ingredient not found!')\n",
    "\n",
    "\n",
    "\n",
    "def delete_ingredient():\n",
    "    search_name_value = search_name.get()\n",
    "\n",
    "    is_exist = False\n",
    "    # xoá theo index\n",
    "    for i in range(len(list_ingredient)):\n",
    "        if list_ingredient[i].name == search_name_value:\n",
    "            del list_ingredient[i]\n",
    "            messagebox.showinfo(\"Delete Ingredient\", f'Delete: {search_name_value} successfully')\n",
    "            is_exist = True # đã tìm thấy nguyên liệu trong hệ thống\n",
    "            break\n",
    "\n",
    "    if is_exist == False: # không tìm thấy nguyên liệu nào trong suốt quá trình lặp\n",
    "        messagebox.showinfo(\"Delete Ingredient\", 'Ingredient not found!')\n",
    "\n",
    "def reset():\n",
    "    name.delete(0, END)\n",
    "    quantity.delete(0, END)\n",
    "    search_name.delete(0, END)\n",
    "\n",
    "\n",
    "\n",
    "# setup các widget\n",
    "# tạo ra các label\n",
    "#label name\n",
    "name_lbl = Label(window, text = 'Name: ')\n",
    "quantity_lbl = Label(window, text = 'Quantity: ')\n",
    "\n",
    "# entry\n",
    "name = Entry(window)\n",
    "quantity = Entry(window)\n",
    "search_name = Entry(window)\n",
    "\n",
    "\n",
    "#tạo button\n",
    "add_btn = Button(window, text = 'Add', command= add_ingredient)\n",
    "search_btn = Button(window, text = 'Search', command= search_ingredient)\n",
    "update_btn = Button(window, text = 'Update', command= update_ingredient)\n",
    "delete_btn = Button(window, text = 'Delete', command= delete_ingredient)\n",
    "reset_btn = Button(window, text = 'Reset', command= reset)\n",
    "\n",
    "# dán các widget lên window\n",
    "# dán từ góc trên bên trái trước, trái sang phải, từ trên xuống dưới\n",
    "name_lbl.grid(row =0, column =0)\n",
    "quantity_lbl.grid(row =1, column =0)\n",
    "add_btn.grid(row =2, column =0)\n",
    "search_name.grid(row =3, column =0)\n",
    "\n",
    "name.grid(row =0, column =1)\n",
    "quantity.grid(row =1, column =1)\n",
    "delete_btn.grid(row =2, column =1)\n",
    "search_btn.grid(row =3, column =1)\n",
    "\n",
    "reset_btn.grid(row =2, column =2)\n",
    "update_btn.grid(row =3, column =2)\n",
    "\n",
    "# chạy ứng dụng\n",
    "window.mainloop()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f40f46e4-4f60-4f98-9e54-0fa7d175586e",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
